### PR TITLE
New GriddedData constructor and unit test 

### LINKED ISSUE #21393: GriddedData constructed from vectors
## Reason
I need to create gridded data from a reporter for multiage transfer

## Design
Add new constructor for GriddedData that takes vectors  as arguments.

### PR DESCRIPTION
**Pull request by aaelmeli**
closes #21393

### CODE REVIEW

File: framework/src/utils/GriddedData.C [hunk 1]
```diff
@@ -57,6 +57,77 @@ GriddedData::GriddedData(std::string file_name)
...
+  if (y_axis.size() != 0)
+  {
+    dim += 1;
+    _axes.push_back(1);
+    grid.resize(dim); // add another dimension to the grid
+    grid[dim - 1].resize(0);
+    grid[dim - 1] = y_axis;
+    num_data_pt *= grid[dim - 1].size();
+  }
+  if (z_axis.size() != 0)
```
Comment (lynnmunday) [inline framework/src/utils/GriddedData.C:89]: put the contents of this if statement into a function that will fill the _grid, _axes and _dim member data.  The return value can be num_data_pt for the checking at the end.  Or you can compute num_data_pt at the end from the grid data

File: framework/src/utils/GriddedData.C [hunk 2]
```diff
@@ -57,6 +57,77 @@ GriddedData::GriddedData(std::string file_name)
...
+               " but ",
+               fcn.size(),
+               " function values were given by user");
+  setGrid(grid, fcn);
+}
+
+void
+GriddedData::setGrid(std::vector<std::vector<Real>> grid, std::vector<Real> data)
+{
+  _grid = grid;
```
Comment (lynnmunday) [inline framework/src/utils/GriddedData.C:123]: fill the _grid data directly instead of copying it in, same with _fcn

**Review by lynnmunday — CHANGES_REQUESTED** [review]

File: framework/src/utils/GriddedData.C [hunk 3]
```diff
@@ -57,6 +57,65 @@ GriddedData::GriddedData(std::string file_name)
...
+  if (num_data_pt != _fcn.size())
+    mooseError("According to the provided axes and function values, number of data points is ",
+               num_data_pt,
+               " but ",
+               _fcn.size(),
+               " function values were given by user");
+  // step is useful in evaluateFcn
+  _step.resize(_dim);
+  _step[0] = 1; // this is actually not used
+  for (unsigned int i = 1; i < _dim; ++i)
```
Comment (lynnmunday) [inline framework/src/utils/GriddedData.C:102]: can you compute num_data_pt in this loop and not compute it in the update grid function.

**Review by lynnmunday — CHANGES_REQUESTED** [review]

File: framework/include/utils/GriddedData.h [hunk 4]
```diff
@@ -59,7 +64,7 @@ class GriddedData
    * If axes[i] == 0 then the i_th axis in the grid data corresponds to the x axis in the simulation
    * If axes[i] == 1 then the i_th axis in the grid data corresponds to the y axis in the simulation
    * If axes[i] == 2 then the i_th axis in the grid data corresponds to the z axis in the simulation
-   * If axes[i] == 3 then the i_th axis in the grid data corresponds to the time in the simulation
+   * If axes[i] == 3 then the i_th axis in the grid data corresponds to thse time in the simulation
```
Comment (GiudGiud) [inline framework/include/utils/GriddedData.h:67]:
```suggestion
   * If axes[i] == 3 then the i_th axis in the grid data corresponds to the time in the simulation
```

File: framework/src/utils/GriddedData.C [hunk 5]
```diff
@@ -57,6 +57,65 @@ GriddedData::GriddedData(std::string file_name)
...
+                         std::vector<Real> y_axis,
+                         std::vector<Real> z_axis,
+                         std::vector<Real> t_axis,
+                         std::vector<Real> fcn)
+{
+  unsigned int num_data_pt = 1;
+  int ind = 0;
+  _axes.resize(0);
+  _fcn = fcn;
+  _dim = 0;
```
Comment (GiudGiud) [inline framework/src/utils/GriddedData.C:70]: you can use initializer list syntax and make things const maybe

File: framework/include/utils/GriddedData.h (hunk 4 above)
Comment (aaelmeli) [reply framework/include/utils/GriddedData.h]: Done

File: framework/src/utils/GriddedData.C (hunk 5 above)
Comment (aaelmeli) [reply framework/src/utils/GriddedData.C]: we moved `_fcn` to the constructor, the other member data need to be computed.

File: framework/src/utils/GriddedData.C (hunk 1 above)
Comment (aaelmeli) [reply framework/src/utils/GriddedData.C:80]: Done.

File: framework/src/utils/GriddedData.C [hunk 6]
```diff
@@ -57,6 +57,77 @@ GriddedData::GriddedData(std::string file_name)
   parse(_dim, _axes, _grid, _fcn, _step, file_name);
 }
 
+GriddedData::GriddedData(std::vector<Real> x_axis,
+                         std::vector<Real> y_axis,
+                         std::vector<Real> z_axis,
+                         std::vector<Real> t_axis,
+                         std::vector<Real> fcn)
+{
```
[the remark this answers is not in the document]
Comment (aaelmeli) [reply framework/src/utils/GriddedData.C:66]: Done.

File: framework/src/utils/GriddedData.C (hunk 2 above)
Comment (aaelmeli) [reply framework/src/utils/GriddedData.C]: Done.

**Comment by lynnmunday** [comment]
@dschwen could you review and merge this?

File: framework/src/utils/GriddedData.C [hunk 7]
```diff
@@ -57,6 +57,69 @@ GriddedData::GriddedData(std::string file_name)
...
+    ind = 2;
+    updateGrid(_dim, z_axis, ind);
+  }
+  if (t_axis.size() != 0)
+  {
+    ind = 3;
+    updateGrid(_dim, t_axis, ind);
+  }
+  if (_dim == 0)
+    mooseError("No valid AXIS lines found by GriddedData");
```
[the remark this answers is not in the document]
Comment (aaelmeli) [reply framework/src/utils/GriddedData.C]: Done.

**Review by dschwen — APPROVED** [review]

**Pull request merged**